KAFKA-15275 - Client state machine basic components, states and initial transitions

lianetm opened the pull request:
Initial definition of the core components for maintaining group membership on the client following the new consumer group protocol.

This PR includes:
- Membership management for keeping member state and assignment, based on the heartbeat responses received. 
- Assignor selection logic to support client and server side assignors. 

This only includes the basic initial states and transitions, to be extended as we implement the protocol. 

This is intended to be used from the heartbeat and assignment requests manager that actually build and process the heartbeat and assignment related requests.

lianetm committed:
Basic components & initial transitions

lianetm committed:
Misc changes for membership management

lianetm committed:
Adding tests

lianetm committed:
Updating descriptions

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignorSelection.java:
```diff
@@ -0,0 +1,102 @@
+ * Server assignors include a name of the server assignor selected, ex. uniform, range.
+ */
+public class AssignorSelection {
+    public enum Type { CLIENT, SERVER }
```
I think the type declaration is redundant because we already know which assignor is used by checking their existence.

Also if we use polymorphic constructor, we don't need to pass null's around.

```
   private Optional<String> serverAssignor = Optional.empty();
    private Optional<List<ConsumerGroupHeartbeatRequestData.Assignor>> clientAssignors = Optional.empty();

    private AssignorSelection(List<ConsumerGroupHeartbeatRequestData.Assignor> assignors) {
        this.clientAssignors = Optional.of(assignors);
    }

    private AssignorSelection(String serverAssignor) {
        this.serverAssignor = Optional.of(serverAssignor);
    }
```

lianetm replied:
Regarding the type, I added it explicitly thinking of cases where we might be interested in checking the type without actually retrieving the assignors (ex. validations). For better readability, seems better to be able to do `if (selection.type() == AssignorSelection.Type.CLIENT)` 

Regarding the constructors, having static factory methods here instead of constructors allows to have meaningful names, so clients using the AssignorSelection can explicitly indicate what they want to create (`newClientAssignors` or `newServerAssignor`). We only pass null internally to the private constructor so seemed like a fair price to pay internally in the class, in order to provide a very explicit way to create new instances.

philipnee replied:
why couldn't we do `hasClientAssignor()` ?

Since we use Optional quite liberally - can we make the assignor Optional as well?

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    private void resetMemberIdAndEpoch() {
+        this.memberId = "";
```
do we ever need to reset memberId?

lianetm replied:
This is used in the case when a member receives UNKNOWN_MEMBER_ID or FENCED_MEMBER_EPOCH from the leader. I'm clearing the memberId mainly for consistency, to avoid keeping it in-memory when we know is wrong. 

Just to double check with @dajac, in the case of these 2 errors, the member should rejoin (providing a member epoch 0), but I'm assuming there is no need to pass the previous member ID.

philipnee replied:
i think the kip says we need to rejoin with the same id : `Upon receiving the UNKNOWN_MEMBER_ID or FENCED_MEMBER_EPOCH error, the consumer abandon all its partitions and rejoins with the same member id and the epoch 0.`

lianetm replied:
Yes, I know it's not written like that but we had conversations saying that sending epoch 0 could be enough (that's why I want to double check)

dajac replied:
In general, we want to keep the member id but we have to clear it when receiving UNKNOWN_MEMBER_ID because as you said it is invalid.

philipnee replied:
would be good to clarify that on the KIP! @dajac

dajac replied:
Actually, we could also keep the member id upon receiving UNKNOWN_MEMBER_ID. The coordinator will use it to recreate the new member. I think that it depends on whether we really want the ability to detect restarts of the consumer or not. If we really want to, keeping the member id would be better. What do you think?

lianetm replied:
+1 on keeping it even in the UNKNOWN_MEMBER_ID scenario. Having the ability to differentiate restarts from brand new consumers seems to open the door for smarter behaviour.

philipnee replied:
it is not entirely clear to me when would a client receive the UNKNOWN_MEMBER_ID error - Obviously, it could be caused by client side bug of sending the wrong/invalid memberId.  On the coordinator side, once a member is assigned to an id - I assume this shouldn't happen? Maybe when a group is being recreated or something?

dajac replied:
The server sends an UNKNOWN_MEMBER_ID error when the received member id does not exist and the received member epoch is greater than 0. This would happen if the member is fenced or manually removed from the consumer group.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberState.java:
```diff
@@ -0,0 +1,69 @@
+     * onPartitionsAssigned or onPartitionsRevoked, and then make the new assignment effective.
+     */
+    // TODO: determine if separate state will be needed for assign/revoke (not for now)
+    PROCESSING_ASSIGNMENT,
```
I guess - this is where we send the first heartbeat?

lianetm replied:
This is when the member receives a new assignment in the hearbeat response. It stays in thit state as long as it is processing the assignment (onPartitionsAssigned and onPartitionsRevoked callbacks execution + make the assignment received the `currentAssignment`). 
This is the interesting state we've been talking a lot about, mainly because we will continue to receive heartbeat responses in this state, and we need to be able to process them correctly while processing the assignment.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -0,0 +1,168 @@
+     * @throws IllegalArgumentException If the provided assignor selection is null
+     */
+    public void setAssignorSelection(AssignorSelection assignorSelection) {
+        if (assignorSelection == null) {
```
we can add the null check at the constructor level as we always need an assignor.

lianetm replied:
We do have the validation at the constructor level (reusing this), so we make sure there's always an assignor. 

For the record, I added this function to update the assignor, with the null check, to enable the case where a member migrates from one type of assignor to another.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -0,0 +1,168 @@
+    @Override
+    public String groupInstanceId() {
+        // TODO: review empty vs null instance id
+        return groupInstanceId.orElse(null);
```
maybe return an optional is safer.

lianetm replied:
Updated (and TODO removed)

philipnee reviewed (commented):
I made a quick first pass - left some initial comments

lianetm committed:
Optional groupInstanceId

lianetm commented:
Build completed with 6 unrelated test failures:

Build / JDK 20 and Scala 2.13 / testBumpTransactionalEpoch(String).quorum=kraft – kafka.api.TransactionsTest
Build / JDK 8 and Scala 2.12 / testNonDefaultConnectionCountLimitAndRateLimit() – kafka.network.ConnectionQuotasTest
Build / JDK 11 and Scala 2.13 / testDelegationTokenRequests(String).quorum=kraft – kafka.server.DelegationTokenRequestsTest
Build / JDK 11 and Scala 2.13 / executeTieredStorageTest(String).quorum=kraft – org.apache.kafka.tiered.storage.integration.ReassignReplicaExpandTest
Build / JDK 17 and Scala 2.13 / testRackAwareRangeAssignor() – integration.kafka.server.FetchFromFollowerIntegrationTest
Build / JDK 17 and Scala 2.13 / testBumpTransactionalEpoch(String).quorum=kraft – kafka.api.TransactionsTest

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignorSelection.java:
```diff
@@ -0,0 +1,102 @@
+ * {@link org.apache.kafka.common.message.ConsumerGroupHeartbeatRequestData.Assignor}
+ * Server assignors include a name of the server assignor selected, ex. uniform, range.
+ */
+public class AssignorSelection {
```
In the first version, we won't support client side assignors. Should we just not care about it for now?

lianetm replied:
Sounds good, will leave the selection but supporting SERVER only for now

kirktrue replied:
Not sure we want to remove the class-level comments about client-side assignors. I don't have a strong preference either way.

lianetm replied:
Removed the client assignor info from the comment too, just to make sure it reflects the reality of what the class currently supports

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberState.java:
```diff
@@ -0,0 +1,69 @@
+     * onPartitionsAssigned or onPartitionsRevoked, and then make the new assignment effective.
+     */
+    // TODO: determine if separate state will be needed for assign/revoke (not for now)
+    PROCESSING_ASSIGNMENT,
```
How about Reconciling?

lianetm replied:
Agree, it's in the end representing the state where the member is reconciling the current assignment with a target assignment received, and aligns better with the terms used in the KIP.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberState.java:
```diff
@@ -0,0 +1,69 @@
+    /**
+     * The member failed with an unrecoverable error
+     */
+    FAILED;
```
I think that we should also have a Fenced state. Fenced is recoverable whereas Failed is not.

lianetm replied:
Totally, added the state and its valid transitions. I also updated the doc for both error states to reflect the differences.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManager.java:
```diff
@@ -0,0 +1,73 @@
+     * @param groupState Group state to be used to compute the assignment
+     * @return Group assignment computed using the selected client side assignor
+     */
+    //TODO: fix parameters and return types to represent the group state object as defined in the
+    // ConsumerGroupPrepareAssignmentResponse
+    Object computeAssignment(Object groupState);
```
Are we going to fix this in this PR?

lianetm replied:
Sure, I just intentionally left it like this to double check here: I guess we want to have the `ConsumerGroupPrepareAssignment` API related classes in the same org.apache.kafka.common.requests package where the heartbeat ones are, correct? (not seeing them created already made me doubt). If that's the right place I'll add them and use them here.

lianetm replied:
Seeing this now in sync with the conversation about leaving out client-side assignors for now, maybe better if I remove this unused `computeAssignment` here, and bring it back when we introduce the client-side assignors logic?

kirktrue replied:
Oof! I'm also working on the code for computing the assignment and invoking callbacks.

See #14357 and KAFKA-15276.

dajac replied:
Yeah, I am +1 for not caring about client side assignor for now.

lianetm replied:
Agree, I removed what was related to client side assignors and computing assignment. 

@kirktrue no worries with what you're working on. I do expect that you bring in the logic for invoking callbacks and computing assignment. I had only added here an unimplemented `computeAssignment` to propose delegating that responsibility to the `membershipManager` once you have access to this code. But we can discuss details when we get to the client assignors part. Removed the compute for now since we're not supporting it.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -0,0 +1,167 @@
+public class MembershipManagerImpl implements MembershipManager {
+
+    private final String groupId;
+    private Optional<String> groupInstanceId;
```
nit: This could be final as well, no?

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -0,0 +1,167 @@
+     */
+    private ConsumerGroupHeartbeatResponseData.Assignment currentAssignment;
+
+    /**
+     * List of assignments that the member received from the server but hasn't processed yet
+     */
+    private final List<ConsumerGroupHeartbeatResponseData.Assignment> targetAssignments;
```
Why do we need a list here? The last one is always the authoritative one. All the others are obsolete. I suppose that the issue is that once a reconciliation between A and B has started, it is not possible to stop it if C comes. Therefore, we have to wait until it completes before we can start the next reconciliation cycle. Do I understand it right?

lianetm replied:
Yes, your understanding is right. We need to keep a single target assignment, but since we'll continue to heartbeat while we reconcile, we also need to keep newer assignments that we might receive while reconciling the 1st one (need to keep only the last one though, as you pointed out). 

Maybe it'll make more clear if I keep something like a single `targetAssignment` (potentially in the process of being reconciled), and a single  optional `nextTargetAssignment`. In the end, I don't expect that this list will ever have more than 2 items.

dajac replied:
I would prefer having three attributes personally because it is clearer.

dajac replied:
I think that we need to think a little more about how we will handle this. My concern is that the server side may start timers based on the `nextTargetAssignment` but the client won't look into it before it finishes the reconciliation to the `targetAssignment`.

For instance, imagine the following scenario.
* Current assignment is A and B;
* Server asks to revoke A and starts the rebalance timer (5 minutes by default);
* Client starts the revocation of A. This can take a while;
* Server asks to revoke B and restarts the rebalance timer;
* Client still waits on the revocation of A to complete. When completed, it starts the revocation of B.

In this scenario, if the revocation of A and B takes more than 5 minutes altogether, the member will be kicked out from the consumer group. The server does not know that the consumer is blocked on the revocation of A.

Do we really need to wait until A is fully revoked before we can start the process to revoke B?

lianetm replied:
Very interesting, we've had some conversations about this (without clear ideas yet) so I need to think about this more. In principle I don't see a reason why we couldn't process the 2 revocations above in parallel, given they relate to different partitions. It feels like the right thing to do, to take advantage of the incremental assignment when possible. That being said, it would definitely be a change in how we execute the callbacks. User's callbacks could be executed concurrently for distinct sets of partitions, which is something that does not happen today (they are invoked sequentially in the poll loop)

I will update this initial code to have the clearer var definitions for now: `currentAssignment`, `targetAssignment` and `nextTargetAssignment`. I will continue to think about how to handle assignments/revocations received while processing a previous one, which seems to be where we could consider the trade-offs of starting processing a `nextTargetAssignment` even if there is a `targetAssignment` being processed (when possible)

dajac replied:
> User's callbacks could be executed concurrently for distinct sets of partitions, which is something that does not happen today (they are invoked sequentially in the poll loop)

This does not seem doable to me because it would break the contract of the `poll` method, isn't it?

The alternative that I am considering would be to not let the server compute a new assignment until the previous one is ack'ed. I need to think this though a little more. The advantage is that it simplifies the client side (we don't need nextTargetAssignment). The disadvantage is that it is a bit less incremental.

pranavrth replied:
There is another case - while A and B revocation is in progress, these partitions are reassigned to the same consumer. In this case `MemberEpoch` will be increased on the coordinator side and OffsetCommit will fail for the ongoing revocations. There maybe alot of commits in progress and client will have ordering issues with retrying those commits.

lianetm replied:
Agree that the actual process on the client should be sync to maintain the current contract. If we stick to the idea that we might continue to receive assignments while a previous one is not acked yet, keeping a nextTargetAssignment (latest one received), could be helpful maybe just to keep it handy (since we had to ignore it). If we move to the server sending assignments only when previous one is acked, then that nextTarget could be left out.

lianetm replied:
the server side change you have in mind seems the easiest to reason about of course, but agree the it's a bit less incremental

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -0,0 +1,167 @@
+            targetAssignments.add(response.assignment());
+            transitionTo(MemberState.PROCESSING_ASSIGNMENT);
+        } else {
+            if (response.errorCode() == Errors.FENCED_MEMBER_EPOCH.code() || response.errorCode() == Errors.UNKNOWN_MEMBER_ID.code()) {
+                resetMemberIdAndEpoch();
+                transitionTo(MemberState.UNJOINED);
+            } else if (response.errorCode() == Errors.UNRELEASED_INSTANCE_ID.code()) {
+                transitionTo(MemberState.FAILED);
+            }
```
Do we plan to handle all the other errors somewhere else?

lianetm replied:
They will all be handled here, to make sure we update the state and other info accordingly. 

I had thought of a follow-up PR for that, mainly to make sure I can go over the errors and understand the ones that are fatal (state should transition to FAILED), and the ones that are somehow recoverable (like the fencing scenario). Adding a TODO here for now, will follow up shortly.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -0,0 +1,167 @@
+        transitionTo(MemberState.STABLE);
+    }
+
+    public void onAssignmentProcessFailure(Throwable error) {
+        // TODO: handle failure scenario when the member was not able to process the assignment
+    }
```
This is interesting. Could you elaborate a bit more on it? Is about failures related to the callbacks?

lianetm replied:
Exactly, this will be for updating the state when exceptions occur in the user provided callbacks. I expect that this will be called from the assign/revoke partitions logic, and that it will transition the state to an unrecoverable FAILED. 

Added the transition to FAILED, as I think we know already that's needed (to ensure same behaviour as the legacy consumer, that fails the poll loop if callbacks fail [here](https://github.com/apache/kafka/blob/84c49c6a098986f7da9f08dc53e6a31c13484c5b/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L484)). I left the TODO with detaila to make sure we properly update the rest of the membership info in this scenario.

dajac replied:
Out of curiosity, do you know how we handle this in the current consumer?

lianetm replied:
In the current consumer, exceptions in the execution of the user-provided callbacks are thrown back to the user from the poll loop as `KafkaException`, breaking the poll loop as they are considered unrecoverable.

dajac replied:
I see. I suppose that if the user calls `poll` again after this, the consumer re-tried to join the group, right?

lianetm replied:
Exactly. The user would get the `KafkaExpcetion` (hopefully do something about it), and call poll again, going through the whole join process

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -0,0 +1,167 @@
+    public void onAssignmentProcessSuccess(ConsumerGroupHeartbeatResponseData.Assignment assignment) {
+        currentAssignment = assignment;
+        targetAssignments.remove(assignment);
+        transitionTo(MemberState.STABLE);
```
With the current logic, I think that we would transition to Stable only if there are no pending assignment to be processed. Or do we retransmission to PROCESSING_ASSIGNMENT somewhere else?

lianetm replied:
Good catch, we do need to consider if there is already another target assignment to reconcile, and only transition to STABLE if there is not.

dajac reviewed (commented):
@lianetm Thanks for the PR. I left a few initial comments/questions.

lianetm committed:
FENCED and RECONCILING states changes

lianetm committed:
Optional groupInstanceId, fixes for transitions to RECONCILING & updated comments/docs

lianetm committed:
Adding tests for Reconciling state and assignor selection

lianetm committed:
Reset memberId only on UNKNOWN_MEMBER_ID & misc fixes for transitions

lianetm committed:
Fixes for memberId, fencing and transitions

lianetm committed:
Adding tests for member id and epoch reset on errors

lianetm committed:
Removing client assignors related logic

lianetm committed:
Adding TODO for extending error handling

lianetm commented:
Thanks for the feedback @dajac, all comments addressed.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignorSelection.java:
```diff
@@ -0,0 +1,82 @@
+ * {@link org.apache.kafka.common.message.ConsumerGroupHeartbeatRequestData.Assignor}
+ * Server assignors include a name of the server assignor selected, ex. uniform, range.
+ */
+public class AssignorSelection {
```
`toString()` please?

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberState.java:
```diff
@@ -0,0 +1,82 @@
+import java.util.Arrays;
+import java.util.List;
+
+public enum MemberState {
```
toString() for debug logging, please? 🥺

lianetm replied:
This is a enum so we'll be getting the value already with the default toString implementation. Do you still prefer to add a custom one including something other than the name?

kirktrue reviewed (changes requested):
Thanks @lianetm!

Did we get direction on whether or not leaving `TODO: ...` in a PR was acceptable practice?

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -0,0 +1,212 @@
+    }
+
+    @Override
+    public void updateState(ConsumerGroupHeartbeatResponseData response) {
```
WDYT about a more descriptive naming: like onHeartbeatResponse().  FWIW: in the heartbeat request manager, I also need to parse the error and response in order to update the local request state. (For exponential backoff). So would it make sense to break it into two functions like: onSuccessfulHeartbeatResponse and onErrorHeartbeatResponse.  Let me know what do you think.

philipnee replied:
in the request manager we would probably do:
{
   if (response.errorCode() == NONE) { 
 .. update some local state
 memberstate.onResponse(response) 
} else {
memberstate.onError(response.errorcode); 
}

lianetm replied:
Uhm actually seems that the heartbeat code you have will end up being simpler if we keep this like we have it now, with a single call to a self contained `updateState` that knows how to update the state based on a HB response (successful or failed). It would only require a single call to `memberState.updateState(response);` and that if/else above would be simpler. WDYT? 
Totally ok with renaming to include onHeartbeatResponse as you suggested.

philipnee replied:
the heartbeat code will also need to handle the error code, for example upon receiving NOT_COORDINATOR error, it will need to invalidate the current coordinator node, so it will need to decode the response.  It is fine for heartbeat to pass the response to the response, but i'm just nit picking about doing the same thing twice.

lianetm committed:
Removing compute assignment logic, fixes for target assignment and member id reset

lianetm committed:
Adding test for assignment and updating UNKNOWN_MEMBER_ID ones

lianetm committed:
Adding toString implementation & removing client assignor comments

lianetm commented:
Thanks for the comments @dajac, all addressed. 

There is still a lot to think about regarding how the client will handle the assignments it receives. In order to unblock the integrations that require this state components, I'm leaving here the initial approach for handling assignments, to continue working on it in a follow-up PR if that's OK. So for now this follows the initial idea that the member may continue to receive assignments while still processing one (so uses current, target and nextTarget with the latest). I updated the transitions and added tests for it, but I expect it will change as we continue to work on it (If we ever move towards the idea of the server not sending new assignments until acked this would be simplified, it not it would probably stay similar to how it is now, requiring more thought on the client side to properly handle all the tricky scenarios we've discussed).

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignorSelection.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    public static AssignorSelection defaultAssignor() {
+        // TODO: review default selection
+        return new AssignorSelection(Type.SERVER, "uniform");
```
By default, we should not set any assignor and let the server use the preferred one.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManager.java:
```diff
@@ -0,0 +1,62 @@
+ */
+public interface MembershipManager {
+
+    String groupId();
```
nit: I would be great o add javadoc to all methods for consistency.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -0,0 +1,240 @@
+ */
+public class MembershipManagerImpl implements MembershipManager {
+
+    private final String groupId;
```
ditto about the javadoc.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -0,0 +1,240 @@
+     * Assignment that the member received from the server and successfully processed.
+     */
+    private ConsumerGroupHeartbeatResponseData.Assignment currentAssignment;
+    /**
```
nit: We usually put an empty line before the javadoc. At least, we do so on the server side.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -0,0 +1,240 @@
+    private boolean hasPendingTargetAssignment() {
+        return targetAssignment.isPresent() || nextTargetAssignment.isPresent();
+    }
+
```
nit: Extra empty line.

dajac reviewed (approved):
@lianetm Thanks for the update. I left a few cosmetic nits. Otherwise, LGTM. I think that we can merge the PR as-is to unblock @philipnee and you can address the remaining comments in the following PRs.

lianetm commented:
Thanks @dajac, I will open a follow-up PR addressing the latest suggestions.
